Incremental Java compiler fails with compile errors due to incorrect classpath, while full compilation works

If a Java class with the same fully-qualified name exists in a compile dependency of a module and also in the module itself, the incremental Java compiler picks the definition provided by the dependency. Only if the sources of the class in the module did 
change, the incremental compiler will resolve to the local one.

### Expected Behavior
Incremental compile should always succeed if a full compilation succeeds and vice versa. It should also consistently select the same definition of a class in case of conflicting definitions.

### Current Behavior
If a class that depends on a class with conflicting definitions is changed, the incremental compilation may fail if the two conflicting definitions are not compatible. If the conflicting class in the local module did also change, the incremental compiler passes it as a changed source file to the compiler. Then the compiler resolves any references for this class to the local definition. If the conflicting class did not change, the class files from the previous compilation of the module are appended to the **end** of the classpath. This gives them the lowest priority during resolution. As a result, references will be resolved to the definition exposed by the external dependency. 

A full rebuild after the failed build succeeds.


### Context
We have multiple occurrences of such conflicting classes in a large code base. The developers often have to trigger a gradle build multiple times until it succeeds (especially after changing Git branches).

### Steps to Reproduce 

#### Module base

BaseClass.java:
```java
package org.example;

public class BaseClass {

}
```

#### Module impl

This module depends on the base module.

BaseClass.java
```java
package org.example;

public class BaseClass {

    public void method() {}

}
```

Impl.java
```java
package org.example;

public class Impl extends BaseClass {

    public void repro(Impl i) {
        i.method();
    }
}
```
Archive containing example:
[duplicateclasses.zip](https://github.com/gradle/gradle/files/9111874/duplicateclasses.zip)

Steps to reproduce:
- Do a full build of the project
- Change Impl.java, but not BaseClass.java
- Trigger another build (now using the incremental compiler)
- Build fails with:

```
  duplicateclasses\impl\src\main\java\org\example\Impl.java:6: error: cannot find symbol
        i.method();
         ^
  symbol:   method method()
  location: variable i of type Impl
```


### Your Environment
Gradle 7.5-RC2 (but also observed on older versions)
Java 11

--- a/platforms/jvm/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalCompilationIntegrationTest.groovy
+++ b/platforms/jvm/language-java/src/integTest/groovy/org/gradle/java/compile/incremental/AbstractCrossTaskIncrementalCompilationIntegrationTest.groovy
@@ -18,6 +18,7 @@
 package org.gradle.java.compile.incremental
 
 import org.gradle.integtests.fixtures.CompiledLanguage
+import spock.lang.Issue
 
 abstract class AbstractCrossTaskIncrementalCompilationIntegrationTest extends AbstractCrossTaskIncrementalCompilationSupport {
 
@@ -304,5 +305,28 @@ abstract class AbstractCrossTaskIncrementalCompilationIntegrationTest extends Ab
 
         then:
         impl.recompiledClasses("A")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/21255")
+    def "ensure local classes preferred in incremental compilation"() {
+        source(
+            api: ["class Base {}"],
+            // Dependant relies on specifics that only exist in the Base within impl
+            impl: ["class Base { String reference; }", "class Dependant { String downstream = new Base().reference;}"]
+        )
+
+        when:
+        run("${language.compileTaskName}")
+        impl.recompiledClasses("Base", "Dependant")
+
+        impl.snapshot()
+
+        // Change the contents of Dependant to force recompile. Compile will only
+        // pass if compiling against impl's Base
+        file("impl/src/main/${language.name}/Dependant.${language.name}") << "// Recompile me please"
+        run("${language.compileTaskName}")
+
+        then:
+        impl.recompiledClasses("Dependant")
     }
 }

--- a/platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java
+++ b/platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java
@@ -16,6 +16,7 @@
 
 package org.gradle.api.internal.tasks.compile.incremental.recomp;
 
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import org.gradle.api.file.FileTree;
@@ -30,7 +31,6 @@ import org.gradle.internal.file.Deleter;
 import org.gradle.work.FileChange;
 
 import java.io.File;
-import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.EnumMap;
@@ -119,7 +119,7 @@ abstract class AbstractRecompilationSpecProvider implements RecompilationSpecPro
         }
     }
 
-    private String rebuildClauseForChangedNonSourceFile(FileChange fileChange) {
+    private static String rebuildClauseForChangedNonSourceFile(FileChange fileChange) {
         return String.format("%s '%s' has been %s", "resource", fileChange.getFile().getName(), fileChange.getChangeType().name().toLowerCase(Locale.US));
     }
 
@@ -265,10 +265,12 @@ abstract class AbstractRecompilationSpecProvider implements RecompilationSpecPro
     }
 
     private static void includePreviousCompilationOutputOnClasspath(JavaCompileSpec spec) {
-        List<File> classpath = new ArrayList<>(spec.getCompileClasspath());
+        List<File> originalClasspath = spec.getCompileClasspath();
         File destinationDir = spec.getDestinationDir();
-        classpath.add(destinationDir);
-        spec.setCompileClasspath(classpath);
+        spec.setCompileClasspath(ImmutableList.<File>builderWithExpectedSize(originalClasspath.size() + 1)
+            .add(destinationDir)
+            .addAll(originalClasspath)
+            .build());
     }
 
     @Override

